File: pythonsrc/train.py
import scipy.io as scio
import numpy as np
#Load the matlab file - Mia
training_data=scio.loadmat('../proj1files/trainingdata.mat')
#Weird Dict, one key being "trainingdata" and the value is an np.ndarray
#Extract that out of that
training_data=training_data["trainingdata"]
print(type(training_data))
#X data
X_train=training_data[:,0:78]
print(np.shape(X_train))
#violent crimes.
Y_violent=training_data[:,79]
print(np.shape(Y_violent))
#non violent crime
Y_non_violent=training_data[:,80]
#Insert Ones into X_train for simple linear regression
X_train = np.insert(X_train, 0, np.ones((X_train.shape[0],)), axis=1)
print(np.shape(X_train))
print(X_train)


#Least Squares Starts Here
betaviolent, *_ = np.linalg.lstsq(X_train, Y_violent, rcond=None)      # shape (80,)
betanonviolent, *_ = np.linalg.lstsq(X_train, Y_non_violent, rcond=None) # shape (80,)
print(len(betaviolent))
print(len(betanonviolent))

#Evaluate model performance
#Recall MAPE is mean absolute percentage error 
# Violent crime MAPE
y_pred_violent = X_train @ betaviolent
MAPE_violent = 100 * np.sum(np.abs(y_pred_violent - Y_violent) / Y_violent) / len(Y_violent)
print(f"MAPE for the violent crime rate training set is {MAPE_violent:.1f}")

# Nonviolent crime MAPE
y_pred_nonviolent = X_train @ betanonviolent
MAPE_nonviolent = 100 * np.sum(np.abs(y_pred_nonviolent - Y_non_violent) / Y_non_violent) / len(Y_non_violent)
print(f"MAPE for the nonviolent crime rate training set is {MAPE_nonviolent:.1f}")

File: proj1files/train.m
%
% Sample training script.  This script uses ordinary linear
% regression to build a model of the data.  
%
%
% Read in the data.
%
load trainingdata.mat
%
% Your code goes below this line
% ------------------------------------------------------
%
% Just use a simple least squares linear regression.  This will not
% be optimal for the MAPE measure.
%
% Separate out X and y.
%
X=trainingdata(:,1:79);
yviolent=trainingdata(:,80);
ynonviolent=trainingdata(:,81);
%
% Insert a column of ones for the beta0 coefficients.
%
X=[ones(size(X,1),1) X];
%
% Find the least squares solutions.
%
betaviolent=X\yviolent;
betanonviolent=X\ynonviolent;
%
% Save the coefficients for later use. The "-V4" option saves a
% file that can be read by Octave as well as MATLAB. 
%
save -V4 beta.mat betaviolent betanonviolent
%
% Output the performance on the training data.
%
MAPEviolent=100*sum(abs(X*betaviolent-yviolent)./yviolent)/length(yviolent);
fprintf(['MAPE for the violent crime rate training set is %.1f\n'],MAPEviolent);
MAPEnonviolent=100*sum(abs(X*betanonviolent-ynonviolent)./ynonviolent)/ length(ynonviolent);
fprintf(['MAPE for the nonviolent crime rate training set is %.1f\n'],MAPEnonviolent);


